crates/matrix-sdk-ui/src/timeline/controller/mod.rs

```diff
@@ -345,6 +345,29 @@ impl<P: RoomDataProvider> TimelineController<P> {
         }
     }
 
+    /// Listens to encryption state changes for the room in
+    /// [`matrix_sdk_base::RoomInfo`] and applies the new value to the
+    /// existing timeline items.
+    pub async fn handle_encryption_state_changes(&self) {
+        let mut room_info = self.room_data_provider.room_info();
+        while let Some(info) = room_info.next().await {
+            let mut changed = false;
+            if let Ok(mut is_room_encrypted) =
+                self.state.read().await.meta.is_room_encrypted.write()
+            {
+                let is_encrypted_now = info.is_encrypted();
+                if is_room_encrypted.is_none() || is_room_encrypted.unwrap() != is_encrypted_now {
+                    changed = true;
+                    *is_room_encrypted = Some(is_encrypted_now);
+                }
+            }
+            if changed {
+                let mut state = self.state.write().await;
```

Comment:
Won't this deadlock? `self.state.read()`, while it doesn't have a binding, will only be dropped after the end of the block. Trying to acquire the write lock won't eve succeed since a read lock is held till the end of the block.

Reply:
The tests pass and they do run this code, so it must be working, although I don't really like how this code looks and I've been trying to improve it, but I'm not sure how to achieve that. Should I keep a reference, then manually drop it, like I did somewhere else?

Reply:
Ok, I think I found a better way to handle this.

Reply:
The way you handled it would have been my recommendation as well.